feat: integrate forum headlines preview and unify homepage styles

fzoonoise opened the pull request:
This update introduces a new `ForumHeadlinesPreview` section on the homepage, consolidates layout styles across components, and improves visual consistency through shared style extraction and font unification.

## Additions  
- New `ForumHeadlinesPreview` component for latest forum headlines display (`forum-headlines-preview.js`)  
- `fetchForumHeadlines` function with JSON fallback → GQL as backup (`utils/api/forum-headlines.js`)  
- New GraphQL query `fetchLatestPublishedExternals`  
- Shared styled component `IndexTitle` for section headers  
- Shared layout wrapper `FullWidthLayout` and style `FullWidthLayoutStyle`  
- New theme config `layout.js` with breakpoint-based `maxWidth`  
- `URL_STATIC_DAILY_COLUMN_HEADLINES` constant for static JSON source  
- Inject `forumHeadlines` into index page `getServerSideProps` 
- `logGenericError` helper in `utils/log/shared.js`
- `processSettledResult` utility to replace `handleAxiosResponse`/`handleGqlResponse` 

## Removals  
- Deprecated `handleAxiosResponse` and `handleGqlResponse` (deleted `utils/response-handle.js`)

## Changes  
- `PromoVideoList`, `LatestNews`, and `index.js` updated to use `IndexTitle`  
- `PromoVideoList` updated to use `FullWidthLayoutStyle` for layout consistency  
- Project-wide default font-family updated to prioritize `'PingFang TC'`, with `'Noto Sans TC'` fallback  
- `ContentWrapper` and inner layout adjusted to handle mobile/tablet breakpoints more precisely  
- Integrated new `theme.layout.maxWidth` into layout styling for responsive consistency  
- `forum-headlines` now uses `publishedDate` exclusively, removing outdated `updatedAt` check  
- Unified post-fetch handling using `processSettledResult` across all pages with `Promise.allSettled`
- Adjust closing brace positions across multiple files to enforce proper block scoping

## Testing  
- Verified `ForumHeadlinesPreview` renders correctly with valid forum data  
- Confirmed ellipsis works for 3-line headline limit across breakpoints  
- Checked fallback to GraphQL when static JSON fails  
- Verified layout consistency after style extractions  
- Ensured no regressions in `PromoVideoList` and `LatestNews` sections  

## Screenshots  
- N/A  

## Todos  
- N/A  

## Task Links  
[【鏡週刊】新增鏡報論壇文章區塊 - Asana](https://app.asana.com/1/614399484723017/project/1210724947067404/task/1210737680990911?focus=true)

Commit:
feat(forum-headlines-preview ): add forum-headlines-preview page and shared layout components

- Add ForumHeadlinesPreview page component
- Add ExternalLayout shared layout component
- Add theme/layout for shared layout styles
- Add fetchForumHeadlines in api/forum-headlines.js
- Add fetchLatestPublishedExternals GraphQL query
- Use fetchForumHeadlines in index page to load data
- Fix grid gap issue in latest-news section

Commit:
refactor(layout): extract index-title shared style component and update external-layout implementation

Commit:
refactor: prioritize using API data from daily-column.json

# Conflicts:
#	packages/mirror-media-next/config/index.mjs

Commit:
refactor(external-layout ): extract ExternalLayout styles for reuse in other components

Commit:
chore: unify default project font and adjust forum-headlines-preview formatting

Commit:
refactor(forum-headlines): extract date validation and clarify timestamp comparison

Commit:
fix: resolve ellipsis overflow issue on tablet and mobile when exceeding three lines

Commit:
feat(headline): add API slug param to determine link target and update to Link component

Commit:
refactor(layout): rename ExternalLayout to FullWidthLayout and update IndexTitle to link to dailycolumn page

v61265 commented on packages/mirror-media-next/components/index/forum-headlines-preview.js:
```diff
@@ -0,0 +1,200 @@
+        <Link href={`./externals/dailycolumn`}>
+          <IndexTitle color="white">鏡報論壇</IndexTitle>
+        </Link>
+        <Ol>
```
[nit]
如果只是大小寫差異的話會比較容易搞混，如果是我會命名 `StyledOl` 或 `ListWrapper` 。

fzoonoise replied:
已更新增加 Styled 前綴字，連同上一個任務 PromoVideoList 相關區塊也更新。

v61265 commented on packages/mirror-media-next/components/index/forum-headlines-preview.js:
```diff
@@ -0,0 +1,200 @@
+
+            return (
+              <Li key={item.id}>
+                <Link href={`./external/${item.slug}`}>
```
這裡 `.` 是多打的嗎？

fzoonoise replied:
的確是用絕對路徑比較好，已更新。

v61265 replied:
你要說的是相對路徑嗎 XDDDDD

fzoonoise replied:
沒有點是絕對路徑吧 (? 從 root 開始延續
是你說的沒有點比較好呀。目前更新版本是沒有點的。

v61265 commented on packages/mirror-media-next/components/index/forum-headlines-preview.js:
```diff
@@ -0,0 +1,200 @@
+            const { date, time } = formatUtcToDateTime(displayTime)
+
+            return (
+              <Li key={item.id}>
```
[nit]
一樣會建議不要取和 tag 拼法相同的 component 名稱，同樣邏輯我可能會叫 `ForumHeadlineItem` 或至少 `StyledLi` 對日後協作者可能比較友善。

fzoonoise replied:
已更新增加 Styled 前綴字，連同上一個任務 PromoVideoList 相關區塊也更新。

v61265 commented on packages/mirror-media-next/pages/index.js:
```diff
@@ -60,10 +62,10 @@ const GA_UTM_EDITOR_CHOICES = 'utm_source=mmweb&utm_medium=editorchoice'
 const IndexContainer = styled.main`
   background-color: rgba(255, 255, 255, 1);
-  max-width: 596px;
+  max-width: ${({ theme }) => theme.layout.maxWidth.md};
```
主動優化讚讚

v61265 commented on packages/mirror-media-next/utils/api/forum-headlines.js:
```diff
@@ -0,0 +1,33 @@
+ * Fetch promote videos data from JSON URL, fallback to GQL API if JSON fails.
+ */
+export async function fetchForumHeadlines() {
+  const jsonData = await axios({
```
[nit]
雖然我也都亂寫，但正宗的寫法其實是 `async/await` 和 `then/catch` 兩者擇一啦（因為如何處理 promise 的邏輯相同），如果這裡要用 await 的話建議使用 `try 和 catch` 比較一致。

fzoonoise replied:
選擇 .catch(() => null) + if (jsonData) 的寫法，是為了避免回傳 Promise<AxiosResponse | ApolloQueryResult> 導致 PromiseSettledResult<ApolloQueryResult> 報錯的型別不一致。

若改用 try...catch 寫法，會因為 return 路徑的控制變嚴格，導致 TypeScript 推斷出更保守的 union type，需額外轉型才可用於 handleGqlResponse，這會引入不必要的型別複雜度。

我有稍微看了一下因應 type 調整後的可能改法，好像太複雜了，想說簡略一點，雖然 coding style 上好像稍微不順，但堪用。

```
  try {
    const jsonRes = await axios({
      method: 'get',
      url: URL_STATIC_DAILY_COLUMN_HEADLINES,
      timeout: API_TIMEOUT,
    })
    return jsonRes
  } catch (err) {
    console.error(
      'Failed to fetch JSON of URL_STATIC_DAILY_COLUMN_HEADLINES, falling back to GQL:',
      JSON.stringify(err)
    )
  }
```

在 index.js 的 handleGqlResponse 會有以下 error:

Argument of type 'PromiseSettledResult<AxiosResponse<any, any> | ApolloQueryResult<any>>' is not assignable to parameter of type 'PromiseSettledResult<ApolloQueryResult<any>>'.
  Type 'PromiseFulfilledResult<AxiosResponse<any, any> | ApolloQueryResult<any>>' is not assignable to type 'PromiseSettledResult<ApolloQueryResult<any>>'.
    Type 'PromiseFulfilledResult<AxiosResponse<any, any> | ApolloQueryResult<any>>' is not assignable to type 'PromiseFulfilledResult<ApolloQueryResult<any>>'.
      Type 'AxiosResponse<any, any> | ApolloQueryResult<any>' is not assignable to type 'ApolloQueryResult<any>'.
        Type 'AxiosResponse<any, any>' is missing the following properties from type 'ApolloQueryResult<any>': loading, networkStatusts(2345)

fzoonoise replied:
經討論後，原定要調整 mapping 所有 return 的值，但優先重構引用的 fetch 後處理 promise 和 logError 的 function (handleAxiosResponse/handleGqlResponse => processSettledResult)，已解決 type 問題，且仍可以保留 async/await + try...catch  寫法。 已全域更新相關內容。
mapping 部分待有需求再更新，避免重構過多內容或後續需較多的設定。

已更新。

v61265 replied:
ok 啦給過，確實如果要重構整個專案的話會太大

v61265 commented on packages/mirror-media-next/utils/api/forum-headlines.js:
```diff
@@ -0,0 +1,33 @@
+    url: URL_STATIC_DAILY_COLUMN_HEADLINES,
+    timeout: API_TIMEOUT,
+  }).catch((err) => {
+    console.error('Failed to fetch JSON, falling back to GQL:', err)
```
[nit]
錯誤處理的部分，以下按照麻煩程度列出來，你就照自己覺得方便的做就好：

1.
`console.error('Failed to fetch JSON of URL_STATIC_DAILY_COLUMN_HEADLINES, falling back to GQL:', JSON.stringify(err))`
1.1. 錯誤訊息要不要標注在哪個 api，之後再後端檢查比較方便。
1.2. 另外如果 error 直接印的話在後台的 log 會被分成一排一排的有點難讀，至少可以加一個 `JSON.strinfy` 這種。

2. 
或是比較完整的這種：https://github.com/mirror-media/Adam/blob/2c6fe14e5c973acf397c36e82a777ceab717b649/packages/mirror-media-next/pages/slot.js#L48

fzoonoise replied:
已更新，加入 SON.stringify 和標記 API 名稱，連同上一個任務 fetchPromoteVideosList 也更新。

v61265 commented on packages/mirror-media-next/components/index/forum-headlines-preview.js:
```diff
@@ -0,0 +1,200 @@
+ * @param {ExternalHeadline[]} props.forumHeadlines - Array of latest forum headlines
+ */
+
+export default function ForumHeadlinesPreview({ forumHeadlines }) {
```
[nit]
為了防止之後要用的人忘記傳值進來，可以加上預設值防呆。
```suggestion
export default function ForumHeadlinesPreview({ forumHeadlines = [] }) {
```

fzoonoise replied:
已更新，連同上一個任務 PromoVideoList 相關區塊也更新。

v61265 commented on packages/mirror-media-next/components/index/forum-headlines-preview.js:
```diff
@@ -0,0 +1,200 @@
+          {forumHeadlines.map((item) => {
+            if (!item.title) return null
+
+            const displayTime = getDisplayTime(
```
這種需求還是第一次見到，看要不要麻煩 PM 加入 spec 中？

fzoonoise replied:
`if (!item.title) return null`

這些是我自己加的防呆耶，想說要是沒有標題的話，會有點尷尬。另一個是說有更新時間的話就顯示更新時間，請問你說要加入 spec 是指？

v61265 replied:
抱歉我是指 Ln.179 的 `getDisplayTime` 看起來是「update 和 create 的時間選最新的」這個邏輯。還是我誤會了？

fzoonoise replied:
對，沒錯，是 pm 要的，要請他加入 spec 嗎？ 但依據每個的資料不同，好像也不一定是這樣的規則。

fzoonoise replied:
經共同討論， PM 後更改為只以建立時間為主，不額外作判斷，已更新。

fzoonoise committed:
refactor: prefix styled HTML tags with Styled, use absolute import for Link, and stringify error in console.log

fzoonoise committed:
refactor: remove link from IndexTitle and update external Headline links to open in new tab

Commit:
feat(api): add handleSettledResponse & logGenericError, refactor fetch methods to try/catch

- Added `handleSettledResponse` as a unified PromiseSettledResult handler
  with transport-aware logging for ApolloError, AxiosError, and generic errors.
- Added `logGenericError` to handle non-Apollo, non-Axios errors with structured logging.
- Updated `fetchForumHeadlines` and `fetchPromoteVideosList` to use try/catch
  for clearer error handling and to avoid PromiseSettledResult type mismatches.
- Adjusted error logging flow to call the appropriate logger based on error type
  instead of relying on separate handleAxiosResponse/handleGqlResponse.
- Ensured no type errors occur when handling mixed Axios/GQL responses.

Commit:
refactor(utils): replace handleAxiosResponse/handleGqlResponse with processSettledResult and unify post-fetch handling

- Introduce `processSettledResult(...)` to process Promise.allSettled() results with transport-aware logging
  (ApolloError → logGqlError, AxiosError → logAxiosError, otherwise → logGenericError).
- Remove `handleAxiosResponse` and `handleGqlResponse`; update imports/usages to call `processSettledResult`
  at all post-fetch call sites.
- Align JSDoc: @template T,V with `PromiseSettledResult<T>` and `(value: T | undefined) => V` mapper signature.
- Maintain stable return shapes by mapping fulfilled values and invoking the mapper with `undefined` on rejection.
- Add inline EN/ZH comments explaining why an explicit `'rejected'` branch is unnecessary (closed union) and the
  routing order (Apollo → Axios → Generic).

BREAKING CHANGE: Direct imports of `handleAxiosResponse`/`handleGqlResponse` are removed. Use `processSettledResult` instead.

# Conflicts:
#	packages/mirror-media-next/pages/section/column.js

Commit:
style: adjust closing brace position to ensure correct block scope execution

Commit:
feat(forum-headlines): remove updatedAt check to always use publishedDate; add explicit return type to API fetch function

Ask ChatGPT

Commit:
fix(promo-video-list): improve preview image resolution when width < 360px by updating width and related styles

# Conflicts:
#	packages/mirror-media-next/components/index/promo-video-list.js

Commit:
fix: align styled component naming after conflict resolution

v61265 reviewed (approved):
LGTM，記得 merge 前解一下 conflict

fzoonoise commented:
> LGTM，記得 merge 前解一下 conflict

有的，早上已經解完。感謝~

fzoonoise committed:
Merge branch 'dev' into dev
